File: Library/DSA Reference/Data Structures/4) Maps/maps.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    ============================================================
    MAPS CHEAT SHEET — C++20
    ============================================================

    CORE / MUST KNOW
    ----------------
    map<K,V>:
        unique keys + sorted by key
        search / insert / erase: O(log n)

    unordered_map<K,V>:
        unique keys + no ordering guarantee
        search / insert / erase: O(1) average, O(n) worst

    multimap<K,V>:
        duplicate keys + sorted by key
        search / insert: O(log n)
        key-based count/erase can include all matching entries

    Python dict:
        unique keys + insertion order preserved
        lookup / insert / delete: O(1) average, O(n) worst

    EXTRAS / REFERENCE
    ------------------
    lower_bound, upper_bound, equal_range, erase-while-iterating,
    custom comparators, hash-policy APIs, merge/extract,
    OrderedDict, defaultdict, Counter-like patterns, etc.
*/

int main() {

    // ============================================================
    // CORE — MUST KNOW
    // ============================================================

    // ------------------------------------------------------------
    // 1. INITIALISATION
    // ------------------------------------------------------------

    map<string, int> mp;
    unordered_map<string, int> unmp;
    multimap<string, int> mmp;

    map<int, string> myMap = {
        {10, "Alice"},
        {20, "Bob"},
        {30, "Charlie"}
    };

    unordered_map<string, int> um = {
        {"A", 1},
        {"B", 2},
        {"C", 3}
    };

    // ------------------------------------------------------------
    // 2. SIZE / EMPTY
    // ------------------------------------------------------------

    int sz = mp.size();             // O(1)
    bool isEmpty = mp.empty();      // O(1)

    cout << "Size: " << sz << '\n';
    cout << "Empty: " << isEmpty << '\n';

    // ------------------------------------------------------------
    // 3. INSERT / UPDATE
    // ------------------------------------------------------------

    // operator[]:
    mp["R"] = 1;
    mp["I"] = 2;
    mp["S"] = 3;
    mp["H"] = 4;
    mp["A"] = 5;
    mp["B"] = 6;

    mp["H"] = 7;                    // Updates existing key

    /*
        IMPORTANT:
        map[key] inserts key with a default-constructed value if the
        key does not already exist.

        So for map<string,int>:
            mp["X"]  -> creates {"X", 0} if X is absent.
    */

    // insert(): does NOT overwrite an existing key
    auto [itInsert, inserted] = mp.insert({"R", 100});
    cout << "Inserted new R? " << boolalpha << inserted << '\n';

    // insert_or_assign(): C++17
    mp.insert_or_assign("R", 100);

    // try_emplace(): C++17
    mp.try_emplace("NEW", 50);

    // unordered_map uses the same key APIs:
    unmp["A"] = 10;
    unmp["B"] = 20;
    unmp["A"] = 99;

    // ------------------------------------------------------------
    // 4. ACCESS
    // ------------------------------------------------------------

    int value1 = mp["R"];            // O(log n), inserts if missing

    auto value2 = mp.at("R");        // O(log n), throws if missing

    // Safe existence check before [] when insertion is unwanted:
    if (mp.find("R") != mp.end())
        cout << mp["R"] << '\n';

    /*
        IMPORTANT:
        `operator[]` is not just a read.
        On a missing key it inserts a default value.
    */

    // ------------------------------------------------------------
    // 5. MEMBERSHIP / FIND
    // ------------------------------------------------------------

    bool present = (mp.find("R") != mp.end());    // O(log n)

    bool presentUS =
        (unmp.find("A") != unmp.end());            // O(1) avg, O(n) worst

    // C++20:
    bool contains = mp.contains("R");              // O(log n)
    bool containsUS = unmp.contains("A");           // O(1) avg, O(n) worst

    cout << "R present: " << present << '\n';

    // ------------------------------------------------------------
    // 6. ERASE
    // ------------------------------------------------------------

    mp.erase("S");                  // O(log n)

    /*
        erase(iterator) returns the iterator immediately after the
        erased element.

        This is extremely useful while erasing during traversal.
    */

    map<int, int> eraseMap = {
        {0,0}, {1,1}, {2,2}, {3,3}, {4,4}
    };

    for (auto it = eraseMap.begin();
         it != eraseMap.end(); ) {

        if (it->first % 2 == 0)
            it = eraseMap.erase(it);
        else
            ++it;
    }

    // Erase by iterator after a safe find:
    auto eraseIt = mp.find("I");
    if (eraseIt != mp.end())
        mp.erase(eraseIt);

    // Erase a range [first,last):
    if (!mp.empty()) {
        auto first = mp.begin();
        auto last = next(first);
        mp.erase(first, last);
    }

    // unordered_map:
    unmp.erase("A");                // O(1) avg, O(n) worst

    // ------------------------------------------------------------
    // 7. TRAVERSAL
    // ------------------------------------------------------------

    cout << "\nmap traversal:\n";

    for (auto el : mp)
        cout << el.first << ' ' << el.second << '\n';

    // Structured bindings:
    for (auto& [key, value] : mp)
        cout << key << " -> " << value << '\n';

    // Important:
    // std::map traversal is sorted by key.
    // unordered_map traversal has no ordering guarantee.

    // ------------------------------------------------------------
    // 8. CLEAR
    // ------------------------------------------------------------

    mp.clear();                     // O(n)

    // ------------------------------------------------------------
    // 9. ORDERED MAP BOUNDS
    // ------------------------------------------------------------

    map<int, string> boundsMap;
    boundsMap[10] = "Alice";
    boundsMap[20] = "Bob";
    boundsMap[30] = "Charlie";

    auto lb = boundsMap.lower_bound(15);  // first key >= 15
    auto ub = boundsMap.upper_bound(15);  // first key > 15

    if (lb != boundsMap.end())
        cout << "Lower bound: "
             << lb->first << " -> "
             << lb->second << '\n';

    if (ub != boundsMap.end())
        cout << "Upper bound: "
             << ub->first << " -> "
             << ub->second << '\n';

    // equal_range:
    auto range = boundsMap.equal_range(20);

    // range.first  == lower_bound(20)
    // range.second == upper_bound(20)

    // ============================================================
    // EXTRAS — REFERENCE
    // ============================================================

    // ------------------------------------------------------------
    // 10. MULTIMAP
    // ------------------------------------------------------------

    multimap<int, string> mmap;

    mmap.insert({1, "apple"});
    mmap.insert({2, "banana"});
    mmap.insert({1, "cherry"});
    mmap.insert({3, "grapes"});
    mmap.insert({2, "kiwi"});

    cout << "\nmultimap traversal:\n";
    for (auto& [key, value] : mmap)
        cout << key << " -> " << value << '\n';

    /*
        Keys are sorted.
        Duplicate keys are allowed.
    */

    // Size:
    int msize = mmap.size();                    // O(1)

    // Count entries with a key:
    size_t cnt3 = mmap.count(3);                // O(log n + k)

    // Find one matching entry:
    auto mit = mmap.find(2);

    if (mit != mmap.end())
        cout << "Found multimap key 2: "
             << mit->second << '\n';

    // Lower / upper bound:
    auto mlb = mmap.lower_bound(2);
    auto mub = mmap.upper_bound(2);

    // All entries with key 2:
    for (auto it = mlb; it != mub; ++it)
        cout << it->first << " -> " << it->second << '\n';

    // Erase ALL entries with key 1:
    mmap.erase(1);

    // Erase ONE entry:
    auto one = mmap.find(2);
    if (one != mmap.end())
        mmap.erase(one);

    // Clear:
    mmap.clear();

    // ------------------------------------------------------------
    // 11. MIN / MAX KEY IN ORDERED MAP
    // ------------------------------------------------------------

    map<int, string> ordered = {
        {10, "Alice"},
        {20, "Bob"},
        {30, "Charlie"}
    };

    if (!ordered.empty()) {
        auto firstPair = *ordered.begin();
        auto lastPair = *ordered.rbegin();

        cout << "Smallest key: "
             << firstPair.first << '\n';

        cout << "Largest key: "
             << lastPair.first << '\n';
    }

    // There is no meaningful smallest/largest key by iteration
    // for unordered_map.

    // ------------------------------------------------------------
    // 12. REVERSE TRAVERSAL
    // ------------------------------------------------------------

    cout << "Reverse map:\n";

    for (auto it = ordered.rbegin();
         it != ordered.rend();
         ++it) {
        cout << it->first
             << " -> "
             << it->second << '\n';
    }

    // ------------------------------------------------------------
    // 13. MAP COPY / MOVE / SWAP
    // ------------------------------------------------------------

    map<int, string> original = ordered;

    map<int, string> copied = original;        // O(n)

    map<int, string> moved = move(original);   // linear in general

    map<int, string> another = {{5, "E"}};

    moved.swap(another);                       // O(1) for map

    // ------------------------------------------------------------
    // 14. CUSTOM ORDERING
    // ------------------------------------------------------------

    map<int, string, greater<int>> descending = {
        {1, "one"},
        {2, "two"},
        {3, "three"}
    };

    cout << "Descending map:\n";

    for (auto& [key, value] : descending)
        cout << key << " -> " << value << '\n';

    // ------------------------------------------------------------
    // 15. CUSTOM COMPARATOR
    // ------------------------------------------------------------

    struct StringLengthCompare {
        bool operator()(const string& a,
                        const string& b) const {

            if (a.size() != b.size())
                return a.size() < b.size();

            return a < b;
        }
    };

    map<string, int, StringLengthCompare> byLength;

    byLength["cat"] = 1;
    byLength["apple"] = 2;
    byLength["dog"] = 3;

    // ------------------------------------------------------------
    // 16. C++20 erase_if
    // ------------------------------------------------------------

    map<int, int> filterMap = {
        {1, 10},
        {2, 20},
        {3, 30},
        {4, 40}
    };

    erase_if(
        filterMap,
        [](const auto& p) {
            return p.first % 2 == 0;
        }
    );

    // ------------------------------------------------------------
    // 17. UNORDERED_MAP HASH POLICY
    // ------------------------------------------------------------

    unordered_map<string, int> hashMap = {
        {"A", 1},
        {"B", 2},
        {"C", 3}
    };

    cout << "Bucket count: "
         << hashMap.bucket_count() << '\n';

    cout << "Load factor: "
         << hashMap.load_factor() << '\n';

    hashMap.reserve(100);
    hashMap.rehash(50);

    /*
        These are advanced hash-table controls.

        For normal DSA:
            unordered_map -> remember average O(1).
        You usually do not need bucket APIs.
    */

    // ------------------------------------------------------------
    // 18. MERGE
    // ------------------------------------------------------------

    map<int, string> left = {
        {1, "one"},
        {2, "two"}
    };

    map<int, string> right = {
        {2, "TWO"},
        {3, "three"}
    };

    left.merge(right);

    /*
        Key 3 moves into left.
        Key 2 cannot move because it already exists in left,
        so it remains in right.
    */

    // ------------------------------------------------------------
    // 19. EXTRACT
    // ------------------------------------------------------------

    map<int, string> extractMap = {
        {1, "one"},
        {2, "two"}
    };

    auto node = extractMap.extract(2);

    if (!node.empty())
        cout << "Extracted: "
             << node.key() << " -> "
             << node.mapped() << '\n';

    // ------------------------------------------------------------
    // 20. USEFUL FREQUENCY PATTERN
    // ------------------------------------------------------------

    vector<string> words = {
        "apple", "banana", "apple", "grape", "banana"
    };

    unordered_map<string, int> freq;

    for (const string& word : words)
        ++freq[word];

    for (const auto& [word, count] : freq)
        cout << word << " -> " << count << '\n';

    /*
        This is one of the most important practical uses of
        unordered_map in DSA:

            frequency[key]++

        Average TC over n inputs: O(n)
    */

    return 0;
}

File: Library/DSA Reference/Data Structures/4) Maps/maps.py
"""
    ============================================================
    MAPS / DICTIONARIES CHEAT SHEET — Python 3.x
    ============================================================

    CORE / MUST KNOW
    ----------------
    Main DSA map type: dict

    dict:
        unique keys
        insertion order preserved
        hash-based
        lookup / insert / delete: O(1) average, O(n) worst

    EXTRAS / REFERENCE
    ------------------
    defaultdict, OrderedDict, Counter, dict views,
    setdefault, update, popitem, comprehensions, etc.
"""

from collections import OrderedDict, defaultdict, Counter


# ================================================================
# CORE — MUST KNOW
# ================================================================

# ------------------------------------------------
# 1. INITIALISATION
# ------------------------------------------------

my_dict = {
    1: "One",
    2: "Two",
    3: "Three"
}

# Constructor:
my_dict2 = dict([
    (1, "One"),
    (2, "Two"),
    (3, "Three")
])

# Dictionary comprehension:
my_dict3 = {
    i: f"Number {i}"
    for i in range(1, 4)
}

# Empty dictionary:
empty = {}

# ------------------------------------------------
# 2. INSERT / UPDATE
# ------------------------------------------------

my_dict = {}

my_dict[1] = "One"              # Insert
my_dict[2] = "Two"              # Insert
my_dict[1] = "Uno"              # Update existing key

print(my_dict)


# ------------------------------------------------
# 3. ACCESS
# ------------------------------------------------

value = my_dict[1]              # Direct access

# If key is missing:
# my_dict[523]                  # KeyError

# Safe access with default:
value = my_dict.get(3, 0)

print(value)

# IMPORTANT:
# get() does NOT insert a missing key.


# ------------------------------------------------
# 4. SIZE / EMPTY
# ------------------------------------------------

n = len(my_dict)                # O(1)
is_empty = not my_dict         # O(1)

print("Size:", n)
print("Empty:", is_empty)


# ------------------------------------------------
# 5. KEY MEMBERSHIP
# ------------------------------------------------

if 2 in my_dict:
    print("Key 2 is present.")

present = 5 in my_dict
missing = 5 not in my_dict

# Average TC: O(1)
# Worst TC: O(n)


# ------------------------------------------------
# 6. REMOVE
# ------------------------------------------------

my_dict = {
    1: "One",
    2: "Two",
    3: "Three"
}

del my_dict[1]                  # Raises KeyError if absent

value = my_dict.pop(2)          # Removes + returns value

print(my_dict)
print(value)

# Safe deletion with default:
value = my_dict.pop(100, None)  # No KeyError


# ------------------------------------------------
# 7. ITERATION
# ------------------------------------------------

my_dict = {
    1: "One",
    2: "Two",
    3: "Three"
}

# Keys:
for key in my_dict:
    print(key)

for key in my_dict.keys():
    print(key)

# Values:
for value in my_dict.values():
    print(value)

# Key-value pairs:
for key, value in my_dict.items():
    print(key, value)


# ------------------------------------------------
# 8. INSERTION ORDER
# ------------------------------------------------

ordered = {
    "apple": 3,
    "banana": 2,
    "cherry": 5
}

for key, value in ordered.items():
    print(key, value)

# Iteration follows insertion order in modern Python dicts.


# ------------------------------------------------
# 9. CLEAR
# ------------------------------------------------

ordered.clear()

# O(n)


# ------------------------------------------------
# 10. UPDATE
# ------------------------------------------------

a = {"A": 1, "B": 2}
b = {"B": 20, "C": 3}

a.update(b)

# Result:
# {"A":1, "B":20, "C":3}

# New keys are inserted; existing values are overwritten.


# ================================================================
# EXTRAS — REFERENCE
# ================================================================

# ------------------------------------------------
# 11. DICTIONARY COMPREHENSION
# ------------------------------------------------

squares = {
    i: i * i
    for i in range(1, 6)
}

even_squares = {
    i: i * i
    for i in range(1, 11)
    if i % 2 == 0
}


# ------------------------------------------------
# 12. setdefault
# ------------------------------------------------

d = {}

d.setdefault("a", 0)
d["a"] += 1

print(d)

# IMPORTANT:
# setdefault(key, default) inserts default if key is absent,
# then returns the value associated with the key.

# It is useful for grouping:
groups = {}

for key, value in [
    ("a", 1),
    ("a", 2),
    ("b", 3)
]:
    groups.setdefault(key, []).append(value)

print(groups)

# TC: O(1) average per dictionary operation.


# ------------------------------------------------
# 13. defaultdict
# ------------------------------------------------

int_dict = defaultdict(int)

int_dict["a"] += 1
int_dict["b"] += 2

print(int_dict)

list_dict = defaultdict(list)

list_dict["a"].append(1)
list_dict["b"].append(2)

print(list_dict)

str_dict = defaultdict(str)

str_dict["a"] += "hello"
str_dict["b"] += "world"

print(str_dict)


# Custom default:
def default_value():
    return "default"

custom_dict = defaultdict(default_value)

print(custom_dict["a"])
print(custom_dict["b"])

# IMPORTANT:
# Accessing a missing defaultdict key CREATES the key
# using the default factory.


# ------------------------------------------------
# 14. OrderedDict
# ------------------------------------------------

ordered_dict = OrderedDict()

ordered_dict["apple"] = 3
ordered_dict["banana"] = 2
ordered_dict["cherry"] = 5

for key, value in ordered_dict.items():
    print(key, value)

# Modern normal dicts already preserve insertion order.
# OrderedDict is mainly useful for its additional ordering APIs.


# ------------------------------------------------
# 15. move_to_end
# ------------------------------------------------

ordered_dict.move_to_end("banana")
print(ordered_dict)

ordered_dict.move_to_end("apple", last=False)
print(ordered_dict)

# Raises KeyError if key is missing.


# ------------------------------------------------
# 16. popitem
# ------------------------------------------------

# Last item:
item = ordered_dict.popitem()
print(item)

# First item:
item = ordered_dict.popitem(last=False)
print(item)

# NOTE:
# Normal Python dict also supports dict.popitem(),
# but it removes the LAST inserted item and has no `last=False` option.


# ------------------------------------------------
# 17. Counter
# ------------------------------------------------

freq = Counter([
    "apple",
    "banana",
    "apple",
    "grape",
    "banana"
])

print(freq)

print("apple:", freq["apple"])

print(freq.most_common(2))

freq.update(["apple", "kiwi"])
freq.subtract(["banana"])

# Counter is extremely useful for frequency counting in DSA.


# ------------------------------------------------
# 18. FREQUENCY MAP — CORE DSA PATTERN
# ------------------------------------------------

values = [1, 2, 2, 3, 3, 3]

freq = {}

for x in values:
    freq[x] = freq.get(x, 0) + 1

print("Frequency:", freq)

# Average TC: O(n)
# SC: O(number of distinct keys)


# ------------------------------------------------
# 19. FREQUENCY WITH defaultdict
# ------------------------------------------------

freq = defaultdict(int)

for x in values:
    freq[x] += 1

print("Frequency:", freq)


# ------------------------------------------------
# 20. GROUPING WITH defaultdict(list)
# ------------------------------------------------

words = [
    ("fruit", "apple"),
    ("fruit", "banana"),
    ("animal", "cat"),
    ("animal", "dog")
]

groups = defaultdict(list)

for category, word in words:
    groups[category].append(word)

print(groups)


# ------------------------------------------------
# 21. DICT VIEWS
# ------------------------------------------------

d = {
    "a": 1,
    "b": 2,
    "c": 3
}

keys = d.keys()
values_view = d.values()
items = d.items()

# These are dynamic view objects, not independent lists.

# Convert if a real list is needed:
keys_list = list(d.keys())
values_list = list(d.values())
items_list = list(d.items())


# ------------------------------------------------
# 22. COPY
# ------------------------------------------------

a = {
    "x": 1,
    "y": 2
}

b = a.copy()
c = dict(a)

# Shallow copies.


# ------------------------------------------------
# 23. MERGE DICTIONARIES
# ------------------------------------------------

a = {"x": 1, "y": 2}
b = {"y": 20, "z": 3}

merged = a | b

# In a key collision, right-hand value wins.

a |= b

# Updates a in place.

# Traditional:
merged2 = {**a, **b}


# ------------------------------------------------
# 24. FILTER A DICTIONARY
# ------------------------------------------------

d = {
    "a": 1,
    "b": 2,
    "c": 3,
    "d": 4
}

filtered = {
    key: value
    for key, value in d.items()
    if value % 2 == 0
}

print(filtered)


# ------------------------------------------------
# 25. SORTED MAP-LIKE VIEWS
# ------------------------------------------------

d = {
    "banana": 2,
    "apple": 3,
    "cherry": 1
}

# Sort by key:
by_key = dict(sorted(d.items()))

# Sort by value:
by_value = dict(
    sorted(d.items(), key=lambda item: item[1])
)

# NOTE:
# A normal dict remains insertion-ordered.
# Sorting creates a new dict whose insertion order is sorted.


# ------------------------------------------------
# 26. FIRST / LAST KEY IN PYTHON DICT
# ------------------------------------------------

d = {
    "a": 1,
    "b": 2,
    "c": 3
}

first_key = next(iter(d))
last_key = next(reversed(d))

# Python dict preserves insertion order, so these mean:
# first inserted key / last inserted key.


# ------------------------------------------------
# 27. SAFE TRAVERSAL WHILE DELETING
# ------------------------------------------------

d = {
    0: "zero",
    1: "one",
    2: "two",
    3: "three"
}

# Do NOT mutate a dict's size directly while iterating:
#
# for key in d:
#     del d[key]      # RuntimeError

# Safe pattern:
for key in list(d):
    if key % 2 == 0:
        del d[key]

print(d)


# ================================================================
# 28. CORE COMPLEXITY MEMORY
# ================================================================

"""
Python dict:

Lookup            O(1) average, O(n) worst
Insert / update   O(1) average, O(n) worst
Delete            O(1) average, O(n) worst
Membership        O(1) average, O(n) worst
len()             O(1)
Iteration         O(n)

Important:
    Keys must be hashable.
    Iteration preserves insertion order.
    Dict has unique keys.
"""


# ------------------------------------------------
# 29. IMPORTANT PYTHON NOTES
# ------------------------------------------------

"""
1. dict is the primary map/hash-map structure in Python.
2. Keys must be hashable.
3. Keys are unique.
4. Assigning an existing key updates its value.
5. dict preserves insertion order in modern Python.
6. d[key] raises KeyError when the key is absent.
7. d.get(key, default) does not insert the key.
8. d.pop(key) removes and returns the value.
9. d.pop(key, default) avoids KeyError if absent.
10. defaultdict creates missing keys when accessed.
11. Counter is useful for frequency counting.
12. OrderedDict is still useful for move_to_end/popitem(last=False)
    even though normal dict already preserves insertion order.
"""
